src/NovelRT/Experimental/Graphics/Vulkan/VulkanGraphicsTexture.cpp

```diff
@@ -0,0 +1,348 @@
+        VkResult mapMemoryResult =
+            vkMapMemory(vulkanDevice, vulkanDeviceMemory, GetOffset(), GetSize(), 0, &pDestination);
+
+        if (mapMemoryResult != VK_SUCCESS)
+        {
+            // TODO: Add real exception.
+            throw std::runtime_error("Failed to map VkMemory to T*! Reason: " + std::to_string(mapMemoryResult));
+        }
+
+        if (readRangeLength != 0)
+        {
+            size_t nonCoherentAtomSize =
+                device->GetAdapter()->GetVulkanPhysicalDeviceProperties().limits.nonCoherentAtomSize;
+            size_t offset = GetOffset() + readRangeOffset;
+            size_t size = (readRangeLength + nonCoherentAtomSize - 1) & ~(nonCoherentAtomSize - 1);
+
+            VkMappedMemoryRange mappedMemoryRange{};
+            mappedMemoryRange.sType = VK_STRUCTURE_TYPE_MAPPED_MEMORY_RANGE;
+            mappedMemoryRange.memory = vulkanDeviceMemory;
+            mappedMemoryRange.offset = offset;
+            mappedMemoryRange.size = size;
+
+            VkResult invalidateResult = vkInvalidateMappedMemoryRanges(vulkanDevice, 1, &mappedMemoryRange);
+
+            if (invalidateResult != VK_SUCCESS)
+            {
+                // TODO: Add real exception.
+                throw std::runtime_error("Failed to invalidate memory ranges to prevent writing! Reason: " +
+                                         std::to_string(invalidateResult));
+            }
+        }
+
+        return reinterpret_cast<uint8_t*>(pDestination) + readRangeOffset;
+    }
+
+    void VulkanGraphicsTexture::Unmap()
+    {
+        std::shared_ptr<VulkanGraphicsDevice> device = GetAllocator()->GetDevice();
+
+        VkDevice vulkanDevice = device->GetVulkanDevice();
+        VkDeviceMemory vulkanDeviceMemory = GetBlock()->GetVulkanDeviceMemory();
+
+        vkUnmapMemory(vulkanDevice, vulkanDeviceMemory);
+    }
+
+    void VulkanGraphicsTexture::UnmapAndWrite()
+    {
+        std::shared_ptr<VulkanGraphicsDevice> device = GetAllocator()->GetDevice();
+
+        VkDevice vulkanDevice = device->GetVulkanDevice();
+        VkDeviceMemory vulkanDeviceMemory = GetBlock()->GetVulkanDeviceMemory();
+
+        size_t nonCoherentAtomSize =
+            device->GetAdapter()->GetVulkanPhysicalDeviceProperties().limits.nonCoherentAtomSize;
+        size_t offset = GetOffset();
+        size_t size = (GetSize() + nonCoherentAtomSize - 1) & ~(nonCoherentAtomSize - 1);
+
+        VkMappedMemoryRange mappedMemoryRange{};
+        mappedMemoryRange.sType = VK_STRUCTURE_TYPE_MAPPED_MEMORY_RANGE;
+        mappedMemoryRange.memory = vulkanDeviceMemory;
+        mappedMemoryRange.offset = offset;
+        mappedMemoryRange.size = size;
+
+        VkResult flushResult = vkFlushMappedMemoryRanges(vulkanDevice, 1, &mappedMemoryRange);
+
+        if (flushResult != VK_SUCCESS)
+        {
+            // TODO: Add real exception.
+            throw std::runtime_error("Failed to flush memory ranges after writing! Reason: " +
```

Comment:
Add a real exception, ye?

Reply:
This will probably be another design ticket or part of an existing one that I mentioned previously.